### PR TITLE
Cooking Nana

### PR DESCRIPTION
**Pull request by NanakoAC**
My much vaunted cooking rework!

Overhauls fryers and ovens to a large degree, cereal/candymaker less so. Refactors all their code, adds a ton of new features

Recipes! All appliances are now compatible with the recipe system, and every recipe has a bitfield of appliances that can make it.  They all default to microwave only, and as yet there aren't any that can be made in multiple appliances. But i added several new  recipes for the fryer, and redistributed a lot of recipes to fryer and oven. Generally most of those related to baking things (pies, cakes, pizza, etc) now use the oven instead of microwave.

Oven and fryer now need preheating. Takes quite a while and uses realistic amounts of power for commercial machinery. Oven has an openable door and rapidly loses heat when its open.

Fryer now uses an actual vat of oil, which can be topped up from spare cooking oil tanks. The efficiency of the fryer varies based on the remaining quantity of oil. spare tanks are found in maintenance or cargo. Also oiltanks can be asploded to spray oil everywhere, i wrote a new reagents proc to spread the contents of a reagent holder all over the local area, just for this (more plans to use this function in future)

All of cooking is generally redesigned to be less about clicking and more about time management and bulk cooking. Microwave can now make large batches of the same recipe, oven+fryer have 4/5 containers that can each cook many things. In general, cooking is SLOW now, you may need to wait several minutes for your fries. But you can concievably cook 20 batches of fries all at once

Adds a coating mechanic and recipes for batter and beer batter, plus a few fryer recipes that use them. Coatings add reagents to the food for extra nutrition, as well as having masked overlays that give it the appearance of being battered.

Combination cooking is reworked to allow using all the ingredients in a cooking container (such as an oven tray), to fuse them all into one giant cake/pizza/etc. the size of the resulting item varies appropriately based on the quantity of stuff used to make it.

Fixes cooking mobs. A deepfried mouse is now converted into a food item that copies the appearance and simulates reagents. uses some code from devouring that i wrote way back when. 

Also adds the mouse burger. We had a sprite for it but it wasnt coded in. thats fixed now. You can stick a fried mouse into a bun

Animated microwave sprite. Since the times are longer, it got boring looking at that static light so now it pulsates

Tweaks several recipes to be a bit more rewarding. Recipe system now has support for producing multiple instances of a result, this is used for several smaller things like cookies and donuts

Basically everything related to cooking is tweaked and fixed up at a low level to make more sense. 
This is a huuuuge content pack ive been working on for weeks, i cant even remember some of the things that are in it.
Look through the code, review. Find the thousands of things ive probably broken. It's a big PR

### CODE REVIEW

**Comment by NanakoAC** [comment]
oh boy conflicts. Ill go resolve those soonish. feel free to review other stuff though, most of the conflicting files are barely related

File: code/_helpers/icons.dm [hunk 1]
```diff
@@ -631,7 +631,6 @@ proc/ColorTone(rgb, tone)
 /*
 Get flat icon by DarkCampainger. As it says on the tin, will return an icon with all the overlays
 as a single icon. Useful for when you want to manipulate an icon via the above as overlays are not normally included.
-The _flatIcons list is a cache for generated icon files.
```
Comment (NanakoAC) [inline code/_helpers/icons.dm]: the flaticons list mentioned here does not exist. so i removed part of this comment

File: code/datums/recipe.dm [hunk 2]
```diff
@@ -1,136 +0,0 @@
-/* * * * * * * * * * * * * * * * * * * * * * * * * *
```
Comment (NanakoAC) [inline code/datums/recipe.dm]: recipe.dm got moved to a more sensible location, next to all the recipe lists

File: code/game/machinery/kitchen/cooking_machines/container.dm [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+			user << span("warning","There's no more space in the [src] for that!")
+			return 0
+
+		if(!user.unEquip(I))
+			return
+		I.forceMove(src)
+		user << span("notice", "You put the [I] into the [src]")
+
+/obj/item/weapon/reagent_containers/cooking_container/verb/empty()
+	set src in view()
```
Comment (Lohikar) [inline code/game/machinery/kitchen/cooking_machines/container.dm]: Shouldn't this be limited to a range of 1, not just `view()`?

File: code/game/machinery/kitchen/cooking_machines/fryer.dm [hunk 4]
```diff
@@ -8,27 +8,165 @@
...
+					else
+						total_our_oil += R.volume
+
+
+	if (total_removed > 0 || total_oil != CI.max_oil)
+		total_oil = min(total_oil, CI.max_oil)
+
+		if (total_our_oil < total_oil)
+			//If we have less than the combined total, then top up from our reservoir
+			var/datum/reagents/buffer = new /datum/reagents(999999999999)
```
Comment (Lohikar) [inline code/game/machinery/kitchen/cooking_machines/fryer.dm]: Why is this so _big_?

File: code/game/turfs/simulated.dm [hunk 5]
```diff
@@ -27,13 +27,9 @@
...
 	if(unwet_task)
-		// Space lube dries a lot.
-		if (wet > 1)
-			unwet_task.trigger_task_in(120 SECONDS)
-		else
-			unwet_task.trigger_task_in(8 SECONDS)
+		unwet_task.trigger_task_in(180 SECONDS)
 	else
-		unwet_task = schedule_task_in(wet > 1 ? 120 SECONDS : 8 SECONDS)
+		unwet_task = schedule_task_in(180 SECONDS)
```
Comment (Lohikar) [inline code/game/turfs/simulated.dm]: Why is this in a cooking overhaul PR?

File: code/modules/reagents/Chemistry-Reagents.dm [hunk 6]
```diff
@@ -115,6 +115,7 @@
 	..()
 	holder = null
 
+
```
Comment (Lohikar) [inline code/modules/reagents/Chemistry-Reagents.dm]: Why?

File: html/changelogs/Nanako-GripperChassis.yml [hunk 7]
```diff
@@ -0,0 +1,39 @@
+################################
```
Comment (Lohikar) [inline html/changelogs/Nanako-GripperChassis.yml]: Why is this here? Merge error?

File: code/game/machinery/kitchen/cooking_machines/fryer.dm (hunk 4 above)
Comment (NanakoAC) [inline code/game/machinery/kitchen/cooking_machines/fryer.dm]: it should be infinity. its a temporary buffer, it needs to hold whatever it needs to hold

File: code/game/turfs/simulated.dm (hunk 5 above)
Comment (NanakoAC) [inline code/game/turfs/simulated.dm]: cooking oil on floors

File: html/changelogs/Nanako-GripperChassis.yml (hunk 7 above)
Comment (NanakoAC) [inline html/changelogs/Nanako-GripperChassis.yml]: yes, seems so, ill do another pass to clean merge issues

File: code/game/turfs/simulated.dm (hunk 5 above)
Comment (Lohikar) [inline code/game/turfs/simulated.dm]: I would like confirmation that this doesn't screw up water drying and other splash effects.

File: code/game/machinery/kitchen/cooking_machines/fryer.dm (hunk 4 above)
Comment (Lohikar) [inline code/game/machinery/kitchen/cooking_machines/fryer.dm]: Why not use `INFINITY` then?

File: code/game/turfs/simulated.dm (hunk 5 above)
Comment (NanakoAC) [inline code/game/turfs/simulated.dm]: it sets all wetness to 180 seconds.

because 8 seconds was dumb

File: code/modules/reagents/Chemistry-Reagents.dm (hunk 6 above)
Comment (NanakoAC) [inline code/modules/reagents/Chemistry-Reagents.dm]: adding in random empty space like this is generally not intentional

Comment (Lohikar) [inline code/modules/reagents/Chemistry-Reagents.dm]: Which is why I pointed it out.

**Comment by NanakoAC** [comment]
all requested changes are done, although im sure reviewing is far from finished

**Comment by NanakoAC** [comment]
please do point out any specific changes you're unsure about, ill explain reasonings!

File: code/game/machinery/kitchen/cooking_machines/_appliance.dm [hunk 8]
```diff
@@ -0,0 +1,652 @@
...
+	// We are trying to cook a grabbed mob.
+	var/obj/item/weapon/grab/G = I
+	if(istype(G))
+
+		if(!can_cook_mobs)
+			user << "<span class='warning'>That's not going to fit.</span>"
+			return 0
+
+		if(!isliving(G.affecting))
+			user << "<span class='warning'>You can't cook that.</span>"
```
Comment (Arrow768) [inline code/game/machinery/kitchen/cooking_machines/_appliance.dm]: Is it intentional that you can cook only living mobs here ?
Wouldnt it make sense to be able to cook both ?
Why would the machine care if some mob is dead or alive if it can fit inside ?

Also if you keep it, it would be nice to have a machine specific message for that, because you you dont "cook" with a mixer.

File: code/game/machinery/kitchen/cooking_machines/_appliance.dm [hunk 9]
```diff
@@ -0,0 +1,652 @@
...
+		usr << string
+	else
+		usr << span("notice","It is empty.")
+
+/obj/machinery/appliance/proc/report_progress(var/datum/cooking_item/CI)
+	if (!CI || !CI.max_cookwork)
+		return null
+
+	if (!CI.cookwork)
+		return "It is cold."
```
Comment (Arrow768) [inline code/game/machinery/kitchen/cooking_machines/_appliance.dm]: Is "It is cold" correct for every appliance (including mixers) ?
Shouldnt it be a machine specific message ?

File: code/game/machinery/kitchen/cooking_machines/_appliance.dm [hunk 10]
```diff
@@ -0,0 +1,652 @@
...
+	// We're trying to cook something else. Check if it's valid.
+	var/obj/item/weapon/reagent_containers/food/snacks/check = I
+	if(istype(check) && islist(check.cooked) && (cook_type in check.cooked))
+		user << "<span class='warning'>\The [check] has already been [cook_type].</span>"
+		return 0
+	else if(istype(check, /obj/item/weapon/reagent_containers/glass))
+		user << "<span class='warning'>That would probably break [src].</span>"
+		return 0
+	else if(istype(check, /obj/item/weapon/disk/nuclear))
+		user << "Central Command would kill you if you [cook_type] that."
```
Comment (Arrow768) [inline code/game/machinery/kitchen/cooking_machines/_appliance.dm]: If you are a [insert antag here] you dont care if centcom would kill you.
Maybe use a different message ?
"The machine buzzes and ejects the [cook_type]"

File: code/game/machinery/kitchen/cooking_machines/_cooker.dm [hunk 11]
```diff
@@ -1,236 +1,145 @@
...
+			//We're above optimal, efficiency goes down as we pass too much over it
+			if (temp_scale >= 2)
+				temp_scale = 0
+			else
+				temp_scale = 1 - (temp_scale - 1)
+
+
+	cooking_power = optimal_power * temp_scale
+
+	//cooking_power *= 1000 //This is for debugging, must remove
```
Comment (Arrow768) [inline code/game/machinery/kitchen/cooking_machines/_cooker.dm]: Must remove ?

Also above you are setting temp_scale to 0.
Then checking if the temperature is > than the min temp
If this is not the case, temp_scale remains at 0 and no power is used ?
--> cooking power remains 0 below 353K

File: code/game/machinery/kitchen/cooking_machines/_cooker.dm [hunk 12]
```diff
@@ -1,236 +1,145 @@
...
-	change_product_appearance(result)
-
-	// Update strings.
-	change_product_strings(result)
-
-	// Set cooked data.
-	var/obj/item/weapon/reagent_containers/food/snacks/food_item = cooking_obj
-	if(istype(food_item) && islist(food_item.cooked))
-		result.cooked = food_item.cooked.Copy()
+/obj/machinery/appliance/cooker
```
Comment (Arrow768) [inline code/game/machinery/kitchen/cooking_machines/_cooker.dm]: Your heating code seems quite complicated.
When you can achieve a minimum temperature with something as simple as a bi-metal switch that turns off as soon as a certain minimum temperature is reached.
(Full power until the min temp; Often used in smaller melting ovens, tea water heaters and simple ovens).

File: code/game/machinery/kitchen/cooking_machines/_appliance.dm (hunk 8 above)
Comment (NanakoAC) [inline code/game/machinery/kitchen/cooking_machines/_appliance.dm]: you misunderstood this. 
living mob is any /mob/living

The only mobs that aren't 'living' are abstract mobs. Dview, AI eye, observer, etc. Things that dont qualify as a mob in any practical sense but still use mob code

File: code/game/machinery/kitchen/cooking_machines/_appliance.dm (hunk 10 above)
Comment (NanakoAC) [inline code/game/machinery/kitchen/cooking_machines/_appliance.dm]: ehhh
this is pre-existing code, it was put in there i imagine because the first thing everyone wants to do is deepfry the nuke disk. i didnt really touch this except to move it around.

Its hard to tell what i touched since i renamed files, so it looks like everything is written from scratch

File: code/game/machinery/kitchen/cooking_machines/_appliance.dm (hunk 9 above)
Comment (NanakoAC) [inline code/game/machinery/kitchen/cooking_machines/_appliance.dm]: well its still correct for the cereal maker.

As for candy, i dunno, but im leaning towards heat being involved somehow

File: code/game/machinery/kitchen/cooking_machines/_appliance.dm (hunk 8 above)
Comment (Arrow768) [inline code/game/machinery/kitchen/cooking_machines/_appliance.dm]: Disregard then

File: code/game/machinery/kitchen/cooking_machines/_cooker.dm (hunk 12 above)
Comment (NanakoAC) [inline code/game/machinery/kitchen/cooking_machines/_cooker.dm]: that is how it works.

it runs at a lower power level to partially preserve the heat, and kicks back to full power if it falls too far from the target

File: code/game/machinery/kitchen/cooking_machines/_cooker.dm (hunk 11 above)
Comment (NanakoAC) [inline code/game/machinery/kitchen/cooking_machines/_cooker.dm]: power is still used as long as the temperature is below target. Cooking power is unrelated to power use. It's a measure of how fast it cooks things

File: code/game/machinery/kitchen/cooking_machines/_appliance.dm (hunk 10 above)
Comment (NanakoAC) [inline code/game/machinery/kitchen/cooking_machines/_appliance.dm]: its generally a standard for the nuke disk to be rejected in various scenarios that might destroy it. i'm not even sure why anymore

File: code/game/machinery/kitchen/cooking_machines/_cooker.dm (hunk 12 above)
Comment (NanakoAC) [inline code/game/machinery/kitchen/cooking_machines/_cooker.dm]: I dont think its really too complicated.

Power  usage and resistance are combined each tick to generate a gain in heat. Resistance is basically an arbitrary value used to create a desired time, with a desired input. its a fudge factor.

Once they get high enough it switches to lower power mode, and the heat has a somewhat reduced chance to drop each tick. without power, it drops faster. when it drops too low the full heat turns back on to bring it back within range.

Also, just like real appliances, the preheating comes with a little red light. When its on, its heating. When the light is off, its ready to cook

File: code/game/machinery/kitchen/cooking_machines/container.dm (hunk 3 above)
Comment (NanakoAC) [inline code/game/machinery/kitchen/cooking_machines/container.dm]: this is fixed now

File: code/game/machinery/kitchen/cooking_machines/_mixer.dm [hunk 13]
```diff
@@ -0,0 +1,134 @@
...
+/obj/machinery/appliance/mixer/can_insert(var/obj/item/I, var/mob/user)
+	if (!stat)
+		user << span("warning","You can't add items while \the [src] is running. Wait for it to finish or turn the power off to abort")
+		return 0
+	else
+		return ..()
+
+/obj/machinery/appliance/mixer/finish_cooking(var/datum/cooking_item/CI)
+	..()
+	stat |= POWEROFF
```
Comment (Lohikar) [inline code/game/machinery/kitchen/cooking_machines/_mixer.dm]: Please set `use_power` to `0` if you want to turn off the machine.

File: code/game/machinery/kitchen/cooking_machines/oven.dm [hunk 14]
```diff
@@ -1,23 +1,110 @@
...
+
+	idle_power_usage	= 5700
+	//uses 30% power to stay warm
+	optimal_power = 0.2
+
+	light_x = 2
+	max_contents = 5
+	container_type = /obj/item/weapon/reagent_containers/cooking_container/oven
+
+	stat = POWEROFF//Starts turned off
```
Comment (Lohikar) [inline code/game/machinery/kitchen/cooking_machines/oven.dm]: Set `use_power` to `0` if you want it to be off.

File: code/game/machinery/kitchen/cooking_machines/_mixer.dm [hunk 15]
```diff
@@ -0,0 +1,134 @@
...
+		return
+
+	if (stat & POWEROFF)//Its turned off
+		stat &= ~POWEROFF
+		if (usr)
+			usr.visible_message("[usr] turns the [src] on", "You turn on the [src]")
+			get_cooking_work(CI)
+			use_power = 2
+	else //Its on, turn it off
+		stat |= POWEROFF
```
Comment (Lohikar) [inline code/game/machinery/kitchen/cooking_machines/_mixer.dm]: `use_power`.

File: code/game/machinery/kitchen/cooking_machines/_mixer.dm [hunk 16]
```diff
@@ -0,0 +1,134 @@
...
+
+1. They have a single container which cant be removed. it will eject multiple contents
+2. Items can't be added or removed once the process starts
+3. Items are all placed in the same container when added directly
+4. They do combining mode only. And will always combine the entire contents of the container into an output
+*/
+
+/obj/machinery/appliance/mixer
+	max_contents = 1
+	stat = POWEROFF
```
Comment (Lohikar) [inline code/game/machinery/kitchen/cooking_machines/_mixer.dm]: `use_power`.

File: code/game/machinery/kitchen/cooking_machines/fryer.dm [hunk 17]
```diff
@@ -8,27 +8,165 @@
...
+	//Power used to maintain temperature once it's heated.
+	//Going with 15% of the active power. This is a somewhat arbitrary value
+
+	resistance = 240000
+	//By default, about 15 mins to heat up.
+
+	max_contents = 4
+	container_type = /obj/item/weapon/reagent_containers/cooking_container/fryer
+
+	stat = POWEROFF//Starts turned off
```
Comment (Lohikar) [inline code/game/machinery/kitchen/cooking_machines/fryer.dm]: `use_power`.

**Comment by NanakoAC** [comment]
requested changes done

**Comment by NanakoAC** [comment]
As per discussion of two days ago, reiterating key points:

the machine stat bitflag is used for data storage and logic handling of several likely common machine states. Broken, maintenance mode, EMP, etc. and the poweroff bit represents a power switch flipped on or off.

There's no low level standard for any of these blocking power usage, so its up to the coder to implement turning off power - via the use power variable, in each of these situations. The use power variable does not imply any specific reason for not drawing power, it could be any of the above. And hence it alone is not adequate to convey a state of "the power switch is flipped off"

My code has now properly implemented setting use_power to 0 when the power switch is turned off, and the poweroff bit in stat is used to store that information. If i did not use that, i'd need to add a new bool to store whether or not its turned on, use_power is still not suitable for holding that data.

**Comment by NanakoAC** [comment]
last commit also adds the cooking oil tank to random stock, and tweaks power usage of mixers. those are things i forgot to do earlier
